src/common/hooks/useInfiniteLoading.ts

```diff
@@ -1,58 +1,144 @@
-  const [loadedData, setLoadedData] = useState<T[]>([]);
-  const rerenderingType = useRef<string>('clear');
+  const [{ items, oldItems, nextItemUrl, count, isGettingMore }, itemDispatch] = useReducer(reducer, {
+    ...initialState,
+    nextItemUrl: initialUrl,
+  });
+  const dispatch = useDispatch();
 
-  const { data, error, isLoading, isFetching } = useQuery(url, options);
+  const { data: fetchedItems, isFetching, isLoading, refetch, error } = useQuery(nextItemUrl, options);
 
-  useEffect(() => {
-    const clear = () => {
-      rerenderingType.current = 'clear';
-      setLoadedData([]);
-      setUrl(initialUrl);
-    };
+  const add = useCallback(
+    (newItem: T) => {
+      itemDispatch({ type: 'add', item: newItem });
+    },
+    [itemDispatch],
+  );
 
-    if (data && !isLoading) {
-      setLoadedData(n => [...n, ...data.results]);
+  const clear = useCallback(() => {
+    itemDispatch({ type: 'reset', nextItemUrl: initialUrl });
+    if (resetApiStateFunction) {
+      dispatch(resetApiStateFunction());
     }
+  }, [itemDispatch, initialUrl, dispatch, resetApiStateFunction]);
 
-    return () => {
-      if (rerenderingType.current === 'clear') {
-        clear();
-      }
-      if (rerenderingType.current === 'fetchMore') {
-        rerenderingType.current = 'clear';
-      }
-    };
-  }, [data, isLoading, initialUrl]);
+  const remove = useCallback(
+    (itemToRemove: T) => {
+      itemDispatch({ type: 'remove', item: itemToRemove });
+    },
+    [itemDispatch],
+  );
 
   const hasMore = useMemo(() => {
     if (isLoading || isFetching) return false;
-    return !!data?.links.next;
-  }, [data, isLoading, isFetching]);
+    return !!fetchedItems?.links.next;
+  }, [fetchedItems, isLoading, isFetching]);
 
-  const fetchMore = () => {
-    if (hasMore && data) {
-      rerenderingType.current = 'fetchMore';
-      setUrl(data.links.next);
+  const getMore = useCallback(() => {
+    if (fetchedItems?.links.next && !isFetching) {
+      itemDispatch({ type: 'set-next-item-url', nextItemUrl: fetchedItems.links.next });
+    }
+  }, [itemDispatch, isFetching, fetchedItems]);
+
+  // Clear the items when the user's internet connection is restored
+  useEffect(() => {
+    if (!isLoading && isFetching && !isGettingMore) {
+      clear();
     }
-  };
-
-  return {
-    loadedData,
-    error,
-    isLoading,
-    isFetching,
-    totalCount: data?.meta.count,
-    hasMore,
-    fetchMore,
-  };
+  }, [isLoading, isFetching, isGettingMore, clear]);
+
+  // Append new items that we got from the API to
+  // oldItems list
+  useEffect(() => {
+    itemDispatch({
+      type: 'add-old',
+      items: fetchedItems?.results || [],
+      count: fetchedItems?.meta.count || 0,
+    });
+  }, [fetchedItems]);
+
+  const itemProviderValue = useMemo(() => {
+    const result = {
+      items: [...items, ...oldItems] as T[],
+      count: items.length + count, // Justin: I'm not so sure this is right.
```

Comment:
Hmm yeah, I think count should be tracked within the reducer itself so we don't have to do strange math here.

I'm not so sure this is right either.

Reply:
Yeah. I'll go ahead and move it. 

By the way, the "I'm not so sure this is right." comment was originally from you. That's why I added the "Justin: " to it. So, far, I haven't recognized any issues with the count. So, we could go ahead and remove that comment.

Reply:
Okay. I was able to refactor this. Check out how I refactored the cases in the reducer.